extensions/amp-smartlinks/0.1/linkmate.js

```diff
@@ -0,0 +1,180 @@
+/**
+ * Copyright 2019 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {deepEquals} from '../../../src/json';
+import {dict} from '../../../src/utils/object';
+
+import {ENDPOINTS} from './constants';
+import {TwoStepsResponse} from
+  '../../amp-skimlinks/0.1/link-rewriter/two-steps-response';
+import {getData} from '../../../src/event-helper';
+
+
+export class Linkmate {
+  /**
+   * @param {?../../../src/service/ampdoc-impl.AmpDoc} ampDoc
+   * @param {?../../../src/service/xhr-impl.Xhr} xhr
+   * @param {?Object} linkmateOptions
+   */
+  constructor(ampDoc, xhr, linkmateOptions) {
+    /** @private {?../../../src/service/ampdoc-impl.AmpDoc} */
+    this.ampDoc_ = ampDoc;
+
+    /** @private {?../../../src/service/xhr-impl.Xhr} */
+    this.xhr_ = xhr;
+
+    /** @private {?boolean} */
+    this.requestExclusiveLinks_ = linkmateOptions.exclusiveLinks;
+
+    /** @private {?number} */
+    this.publisherID_ = linkmateOptions.publisherID;
+
+    /** @private {?string} */
+    this.linkAttribute_ = linkmateOptions.linkAttribute;
+
+    /** @private {!Document|!ShadowRoot} */
+    this.rootNode_ = this.ampDoc_.getRootNode();
+
+    /** @private {?Array<!HTMLElement>} */
+    this.anchorList_ = null;
+
+    /** @private {?Array<JsonObject>}*/
+    this.linkmateResponse_ = null;
+  }
+
+  /**
+   * Callback used by LinkRewriter. Whenever there is a change in the anchors
+   * on the page we want make a new API call.
+   * @param {!Array<!HTMLElement>} anchorList
+   * @return {!../../amp-skimlinks/0.1/link-rewriter/two-steps-response.TwoStepsResponse}
+   * @public
+   */
+  runLinkmate(anchorList) {
+    // If we already have an API response and the anchor list has
+    // changed since last API call then map any new anchors to existing
+    // API response
+    let syncMappedLinks = null;
+    if (this.linkmateResponse_ && this.anchorList_ &&
+      !deepEquals(this.anchorList_, anchorList)) {
```

Comment:
const anchorListChanged = this.anchorList_ && !deepEquals(this.anchorList_, anchorList);
    if (anchorListChanged) {
      if (this.linkmateResponse_) {
        syncMappedLinks = this.mapLinks_();
      } else {
         // We don't have an API response...
         const asyncMappedLinks = this.postToLinkmate_(anchorList)
         ...
      }
    } else {
       // If we didn't need to make an API
      this.anchorList_ = anchorList;
      return new TwoStepsResponse(syncMappedLinks, null);
    }

Reply:
@alabiaga

Thanks for the `anchorListChanged` suggestion! It makes the whole block much more readable.

I don't think I can structure the if statements the way you suggested b/c we need an initial check for when there is no API response and no `this.anchorList_`. I could add this as an `else if` on the outside `if` in your example or keep it the way I have now. Which would you prefer?

Reply:
You can just keep your changes but extract out your usage of `this.anchorList_ && !deepEquals(this.anchorList_, anchorList)` to a const for reuse. Thanks